spec/requests/users_controller_spec.rb

```diff
@@ -1,51 +1,2417 @@
+        before { SiteSetting.must_approve_users = true }
+
+        it 'creates a user correctly' do
+          SiteSetting.queue_jobs = true
+
+          post_user
+          expect(response.status).to eq(200)
+
+          expect(JSON.parse(response.body)['active']).to be_falsey
+
+          # should save user_created_message in session
+          expect(session["user_created_message"]).to be_present
+          expect(session[SessionController::ACTIVATE_USER_KEY]).to be_present
+
+          expect(Jobs::SendSystemMessage.jobs.size).to eq(0)
+
+          expect(Jobs::CriticalUserEmail.jobs.size).to eq(1)
+          args = Jobs::CriticalUserEmail.jobs.first["args"].first
+          expect(args["type"]).to eq("signup")
+        end
+      end
+
+      context 'users already exists with given email' do
+        let!(:existing) { Fabricate(:user, email: post_user_params[:email]) }
+
+        it 'returns an error if hide_email_address_taken is disabled' do
+          SiteSetting.hide_email_address_taken = false
+
+          post_user
+          expect(response.status).to eq(200)
+
+          json = JSON.parse(response.body)
+          expect(json['success']).to eq(false)
+          expect(json['message']).to be_present
+        end
+
+        it 'returns success if hide_email_address_taken is enabled' do
+          SiteSetting.hide_email_address_taken = true
+          expect {
+            post_user
+          }.to_not change { User.count }
+
+          expect(response.status).to eq(200)
+          expect(session["user_created_message"]).to be_present
+
+          json = JSON.parse(response.body)
+          expect(json['active']).to be_falsey
+          expect(json['message']).to eq(I18n.t("login.activate_email", email: post_user_params[:email]))
+        end
+      end
+    end
+
+    context "creating as active" do
+      it "won't create the user as active" do
+        post "/u.json", params: post_user_params.merge(active: true)
+        expect(response.status).to eq(200)
+        expect(JSON.parse(response.body)['active']).to be_falsey
+      end
+
+      context "with a regular api key" do
+        let(:user) { Fabricate(:user) }
+        let(:api_key) { Fabricate(:api_key, user: user) }
+
+        it "won't create the user as active with a regular key" do
+          post "/u.json",
+            params: post_user_params.merge(active: true, api_key: api_key.key)
+
+          expect(response.status).to eq(200)
+          expect(JSON.parse(response.body)['active']).to be_falsey
+        end
+      end
+
+      context "with an admin api key" do
+        let(:admin) { Fabricate(:admin) }
+        let(:api_key) { Fabricate(:api_key, user: admin) }
+
+        it "creates the user as active with a regular key" do
+          SiteSetting.queue_jobs = true
+          SiteSetting.send_welcome_message = true
+          SiteSetting.must_approve_users = true
+
+          #Sidekiq::Client.expects(:enqueue).never
```

Comment:
@tgxworld thanks for the feedback! I've updated the PR according to your feedback except for this bit here. I looked online for ways to find the number of queued items and every method I tried reported 0 queued items which is exactly what we need to verify. However, I altered the test so that jobs got queued (e.g. `Jobs::CriticalUserEmail.jobs.size` says 1) but everything I tried still reported 0 queued items.

For now I've added those two lines below:
```ruby
expect(Jobs::CriticalUserEmail.jobs.size).to eq(0)
expect(Jobs::SendSystemMessage.jobs.size).to eq(0)
```

which I think are equivalent to what this test is verifying.

Reply:
I think the two assertions which you've added is good enough :+1: